### PR TITLE
feat: add `codehash` and `storage-root` to `cast`

### LINKED ISSUE #8823: feat(cast): add `cast codehash` and `cast storage-root`
### Component

Cast

### Describe the feature you would like

We currently have commands `cast balance` and `cast nonce` to examine these fields of an account's state. Yet we still lack a `cast codehash` or `cast storage-root`. I propose that these be added to round this out.

Regarding #5467 these would under `account` as subcommands although I'm not sure if they should be added before or after we start breaking the commands out into the different categories. What does everyone think?

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Sabnock01**
## Motivation

Currently, `cast` has the commands `balance` and `nonce` that allows for retrieving the values of these fields of an account's state. However, other fields such as codehash and storage root are not available.

## Solution

This PR introduces two new commands `cast codehash` and `cast storage-root` to allow for these fields retrieval. Resolves #8823

### CODE REVIEW

File: crates/cast/src/lib.rs [hunk 1]
```diff
@@ -494,6 +494,48 @@ where
...
+    /// let provider =
+    ///     ProviderBuilder::<_, _, AnyNetwork>::default().on_builtin("http://localhost:8545").await?;
+    /// let cast = Cast::new(provider);
+    /// let addr = Address::from_str("0x7eD52863829AB99354F3a0503A622e82AcD5F7d3")?;
+    /// let codehash = cast.codehash(addr, None).await?;
+    /// println!("{}", codehash);
+    /// # Ok(())
+    /// # }
+    pub async fn codehash(&self, who: Address, block: Option<BlockId>) -> Result<String> {
+        Ok(self.provider.get_account(who).block_id(block.unwrap_or_default()).await?.code_hash.to_string())
```
Comment (klkvr) [inline crates/cast/src/lib.rs:515]: afaik `eth_getAccount` is pretty new and is not supported by eg alchemy, should we consider falling back to `eth_getProof`?

Comment (Sabnock01) [reply crates/cast/src/lib.rs:515]: Sure!

File: crates/cast/bin/args.rs [hunk 2]
```diff
@@ -83,7 +83,7 @@ pub enum CastSubcommand {
     },
 
     /// Concatenate hex strings.
-    #[command(visible_aliases = &["--concat-hex", "ch"])]
+    #[command(visible_aliases = &["--concat-hex", "cx"])]
```
Comment (mattsse) [inline crates/cast/bin/args.rs:86]: I'd like not to change aliases

File: crates/cast/bin/args.rs [hunk 3]
```diff
@@ -831,6 +831,48 @@ pub enum CastSubcommand {
         rpc: RpcOpts,
     },
 
+    /// Get the codehash for an account.
+    #[command(visible_alias = "ch")]
```
Comment (mattsse) [inline crates/cast/bin/args.rs:835]: imo codehash should be fine or chash

File: crates/cast/src/lib.rs [hunk 4]
```diff
@@ -494,6 +494,72 @@ where
...
+    /// let addr = Address::from_str("0x7eD52863829AB99354F3a0503A622e82AcD5F7d3")?;
+    /// let slots = vec![FixedBytes::from_str("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")?];
+    /// let storage_root = cast.storage_root(addr, slots, None).await?;
+    /// println!("{}", storage_root);
+    /// # Ok(())
+    /// # }
+    pub async fn storage_root(
+        &self,
+        who: Address,
+        slots: Vec<FixedBytes<32>>,
```
Comment (mattsse) [inline crates/cast/src/lib.rs:551]:
```suggestion
        slots: Vec<B256>,
```

File: crates/cast/src/lib.rs [hunk 5]
```diff
@@ -494,6 +494,72 @@ where
...
+    /// let addr = Address::from_str("0x7eD52863829AB99354F3a0503A622e82AcD5F7d3")?;
+    /// let slots = vec![FixedBytes::from_str("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")?];
+    /// let codehash = cast.codehash(addr, slots, None).await?;
+    /// println!("{}", codehash);
+    /// # Ok(())
+    /// # }
+    pub async fn codehash(
+        &self,
+        who: Address,
+        slots: Vec<FixedBytes<32>>,
```
Comment (mattsse) [inline crates/cast/src/lib.rs:518]:
```suggestion
        slots: Vec<B256>,
```

**Review by mattsse — CHANGES_REQUESTED** [review]
some nits

File: crates/cast/bin/args.rs (hunk 3 above)
Comment (Sabnock01) [reply crates/cast/bin/args.rs:835]: Never seen `chash` so I think I'll leave it with no alias

File: crates/cast/src/lib.rs [hunk 6]
```diff
@@ -494,6 +494,72 @@ where
...
+    /// let codehash = cast.codehash(addr, slots, None).await?;
+    /// println!("{}", codehash);
+    /// # Ok(())
+    /// # }
+    pub async fn codehash(
+        &self,
+        who: Address,
+        slots: Vec<B256>,
+        block: Option<BlockId>,
+    ) -> Result<String> {
```
Comment (mattsse) [inline crates/cast/src/lib.rs:520]:
```suggestion
    ) -> Result<B256> {
```

File: crates/cast/src/lib.rs [hunk 7]
```diff
@@ -494,6 +494,72 @@ where
...
+        Ok(self
+            .provider
+            .get_proof(who, slots)
+            .block_id(block.unwrap_or_default())
+            .await?
+            .code_hash
+            .to_string())
```
Comment (mattsse) [inline crates/cast/src/lib.rs:527]: @klkvr should we try get_account first and only if that fails, get_proof?

File: crates/cast/src/lib.rs [hunk 8]
```diff
@@ -494,6 +494,72 @@ where
...
+    /// let storage_root = cast.storage_root(addr, slots, None).await?;
+    /// println!("{}", storage_root);
+    /// # Ok(())
+    /// # }
+    pub async fn storage_root(
+        &self,
+        who: Address,
+        slots: Vec<B256>,
+        block: Option<BlockId>,
+    ) -> Result<String> {
```
Comment (mattsse) [inline crates/cast/src/lib.rs:553]:
```suggestion
    ) -> Result<B256> {
```

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/cast/src/lib.rs (hunk 7 above)
Comment (Sabnock01) [reply crates/cast/src/lib.rs:527]: Could have a `match` to check if `slots` is present, use `get_proof` if it is and `get_account` if it isn't. Similar to `cast balance`

File: crates/cast/src/lib.rs (hunk 8 above)
Comment (DaniPopes) [reply crates/cast/src/lib.rs:553]: functions in this file return string

Comment (mattsse) [reply crates/cast/src/lib.rs:553]: right

**Review by mattsse — APPROVED** [review]

**Pull request merged**